Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+        break;
+    case 'G':
+        pattern_space.append('\n');
+        pattern_space.append(hold_space.string_view());
+        break;
+    case 'h':
+        hold_space = pattern_space;
+        break;
+    case 'H':
+        hold_space.append('\n');
+        hold_space.append(pattern_space.string_view());
+        break;
+    case 'l':
+        print_unambiguous(stdout, pattern_space.string_view());
+        break;
+    case 'n':
+        if (!suppress_default_output)
+            write_pattern_space(stdout, pattern_space);
+        if (input.has_next()) {
+            pattern_space.clear();
+            pattern_space.append(input.next());
+        }
+        break;
+    case 'p':
+        write_pattern_space(stdout, pattern_space);
+        break;
+    case 'P': {
+        auto pattern_sv = pattern_space.string_view();
+        auto newline_position = pattern_sv.find('\n').value_or(pattern_sv.length() - 1);
+        stdout->write(pattern_sv.substring_view(0, newline_position + 1));
+        break;
+    }
+    case 'q':
+        cycle_decision = CycleDecision::Quit;
+        break;
+    case 'x':
+        swap(pattern_space, hold_space);
+        break;
+    case '=':
+        stdout->write(String::formatted("{}\n", input.line_number()));
+        break;
+    case '#':
+        break;
+    default:
+        warnln("Command not implemented: {}", command.function);
+        break;
+    }
+
+    return cycle_decision;
+}
+
+static bool run(Vector<InputFile>& inputs, Script& script, bool suppress_default_output)
+{
+    // TODO: verify all commands are valid
+
+    StringBuilder pattern_space;
+    StringBuilder hold_space;
+    auto stdout = Core::File::standard_output();
+
+    // TODO: extend to multiple input files
+    auto& input = inputs[0];
+    // main loop
+    while (input.has_next()) {
+
+        // TODO: "Reading from input shall be skipped if a <newline> was in the pattern space prior to a D command ending the previous cycle"
+        pattern_space.append(input.next());
+        auto is_last_line = !input.has_next();
+
+        // Turn commands on/off depending on selection. We need
+        for (auto& command : script.commands())
+            command.enable_for(pattern_space.string_view(), input.line_number(), is_last_line);
+
+        // Go, go, go!
+        CycleDecision cycle_decision = CycleDecision::None;
+        for (auto& command : script.commands()) {
+            if (!command.is_enabled())
+                continue;
+            auto command_cycle_decision = apply(command, pattern_space, hold_space, input, suppress_default_output);
+            if (command_cycle_decision == CycleDecision::Next || command_cycle_decision == CycleDecision::Quit) {
+                cycle_decision = command_cycle_decision;
+                break;
+            }
+        }
+
+        if (cycle_decision == CycleDecision::Next) {
+            continue;
+        } else if (cycle_decision == CycleDecision::Quit) {
```

Comment:
No need for `else` after that `continue`.

Reply:
I don't understand the comment. `cycle_decision` can be `None`, `Next` or `Quit`, and in this `if`/`else if` block the three values give the three different, expected behaviors. Not having this `else if` here wouldn't allow me to implement the `Quit` cycle decision.

Reply:
I was suggesting this:
```suggestion
        if (cycle_decision == CycleDecision::Quit) {
```

(i.e. just the "else", not the whole else-if)

Reply:
Ahhh, I see now, sorry for the confusion, changed now.